filters/auth/tokeninfo.go

```diff
@@ -395,3 +429,39 @@ func (f *tokeninfoFilter) Request(ctx filters.FilterContext) {
 }
 
 func (f *tokeninfoFilter) Response(filters.FilterContext) {}
+
+func (f *tokeninfoValidateFilter) Request(ctx filters.FilterContext) {
+	if _, ok := ctx.StateBag()[tokeninfoCacheKey]; ok {
+		return // tokeninfo was already validated by a preceding filter
+	}
+
+	if len(f.config.OptOutAnnotations) > 0 {
+		annotations := annotate.GetAnnotations(ctx)
+		for _, annotation := range f.config.OptOutAnnotations {
+			if _, ok := annotations[annotation]; ok {
+				return // opt-out from validation
+			}
+		}
+	}
+
+	if token, ok := getToken(ctx.Request()); ok {
+		tokeninfo, err := f.client.getTokeninfo(token, ctx)
+		if err == nil { // token is valid
```

Comment:
hmm rather not obvious condition, also the `ok` check instead of `!ok` is not really obvious by the rule of "errors to the right and ok path to the left."

Reply:
Please suggest how to rewrite to keep the same logic.

Reply:
As far as I understood Sandor, something like this:
```go
func (f *tokeninfoValidateFilter) serveOnError(ctx filters.FilterContext) {
	ctx.Serve(&http.Response{
		StatusCode: http.StatusUnauthorized,
		Header: http.Header{
			"Content-Length": []string{strconv.Itoa(len(f.config.UnauthorizedResponse))},
		},
		Body: io.NopCloser(strings.NewReader(f.config.UnauthorizedResponse)),
	})
}
```

```go
	token, ok := getToken(ctx.Request())
	if !ok {
		f.serveOnError(ctx)
		return
	}

	tokeninfo, err := f.client.getTokeninfo(token, ctx)
	if err != nil {
		f.serveOnError(ctx)
		return
	}

	uid, _ := tokeninfo[uidKey].(string)
	authorized(ctx, uid)
	ctx.StateBag()[tokeninfoCacheKey] = tokeninfo
```

Reply:
I do not see how it is better, moreover I see an opportunity to forget one of the early returns.

Reply:
That's how I see Sandor's suggestion:
```go
func (f *tokeninfoValidateFilter) serveOnError(ctx filters.FilterContext) {
	ctx.Serve(&http.Response{
		StatusCode: http.StatusUnauthorized,
		Header: http.Header{
			"Content-Length": []string{strconv.Itoa(len(f.config.UnauthorizedResponse))},
		},
		Body: io.NopCloser(strings.NewReader(f.config.UnauthorizedResponse)),
	})
}
```

```go
token, ok := getToken(ctx.Request())
	if !ok {
		f.serveOnError(ctx)
		return
	}

	tokeninfo, err := f.client.getTokeninfo(token, ctx)
	if err != nil {
		f.serveOnError(ctx)
		return
	}

	uid, _ := tokeninfo[uidKey].(string)
	authorized(ctx, uid)
	ctx.StateBag()[tokeninfoCacheKey] = tokeninfo
	return
```

Reply:
Okay, I think I'm fine with both options.

Reply:
yes @RomanZavodskikh code is what I would prefer.

Reply:
I don't think it is better but I changed to the suggested variant.